gensim/models/fasttext.py

```diff
@@ -1324,3 +1163,490 @@ def save_facebook_model(model, path, encoding="utf-8", lr_update_rate=100, word_
+
+        Parameters
+        ----------
+        fb_vectors : np.array
+            A matrix containing vectors for all the entities, including words
+            and ngrams.  This comes directly from the binary model.
+            The order of the vectors must correspond to the indices in
+            the vocabulary.
+        match_gensim : boolean, optional
+            No longer supported.
+
+        """
+        vocab_words = len(self)
+        assert fb_vectors.shape[0] == vocab_words + self.bucket, 'unexpected number of vectors'
+        assert fb_vectors.shape[1] == self.vector_size, 'unexpected vector dimensionality'
+
+        #
+        # The incoming vectors contain vectors for both words AND
+        # ngrams.  We split them into two separate matrices, because our
+        # implementation treats them differently.
+        #
+        self.vectors_vocab = np.array(fb_vectors[:vocab_words, :])
+        self.vectors_ngrams = np.array(fb_vectors[vocab_words:, :])
+        self.recalc_word_ngram_buckets()
+        self.adjust_vectors()  # calculate composite full-word vectors
+
+    def adjust_vectors(self):
+        """Adjust the vectors for words in the vocabulary.
+
+        The adjustment composes the trained full-word-token vectors with
+        the vectors of the subword ngrams, matching the Facebook reference
+        implementation behavior.
+
+        """
+        if self.bucket == 0:
+            return
+
+        self.vectors = self.vectors_vocab[:].copy()
+        for i, w in enumerate(self.index_to_key):
+            ngram_buckets = self.buckets_word[i]
+            for nh in ngram_buckets:
+                self.vectors[i] += self.vectors_ngrams[nh]
+            self.vectors[i] /= len(ngram_buckets) + 1
+
+    def recalc_word_ngram_buckets(self):
+        """
+        Scans the vocabulary, calculates ngrams and their hashes, and cache the
+        list of ngrams for each known word.
+
+        TODO: evaluate if precaching even necessary, compared to recalculating as needed
+        """
+        if self.bucket == 0:
+            self.buckets_word = [np.array([], dtype=np.uint32)] * len(self.index_to_key)
+            return
+
+        self.buckets_word = [None] * len(self.index_to_key)
+
+        for i, word in enumerate(self.index_to_key):
+            self.buckets_word[i] = np.array(
+                ft_ngram_hashes(word, self.min_n, self.max_n, self.bucket, self.compatible_hash),
+                dtype=np.uint32,
+            )
+
+
+def _pad_random(m, new_rows, rand):
+    """Pad a matrix with additional rows filled with random values."""
+    rows, columns = m.shape
+    low, high = -1.0 / columns, 1.0 / columns
+    suffix = rand.uniform(low, high, (new_rows, columns)).astype(REAL)
+    return vstack([m, suffix])
+
+
+def _rollback_optimization(kv):
+    """Undo the optimization that pruned buckets.
+
+    This unfortunate optimization saves memory and CPU cycles, but breaks
```

Comment:
Better than tweaking language would be to remove all remnants of this misguided deviation from real FastText's implementation decisions and inherent design goals. So once again, in order to avoid this tail of maintenance work for long-ago backward compatibility, that's costing us unbilled time, and leaving cruft in the codebase, but quite possibly benefiting exactly 0 actual users, can we just pick one clear backward-compatibility goal: "4.0 aims to load models that were saved from `gensim-3.8.4`", and jettison the rest?

Reply:
Yes, possibly. When I see something trivial that needs changing during a review, I just change it. Saves effort & mental capacity.

Regardless of more far-reaching redesigns that may come later.

Reply:
The verb-tense of this comment (which predated this PR and I didn't write) was still plausibly correct: the optimization still exists in the older loaded models handled by this method, and until changed (& while still optionally supported!) still breaks compatibility with canonical Facebook FastText & current gensim Fasttext models. So rather than polishing the comments on technical debt code that should be removed as soon as practical, I've just jettisoned the unwise support. The one existing point where the nonstandard hash was explicitly detected – in the loading of `FastTextKeyedVectors` – I've added an error counseling the user to try upgrading their model using gensim-3.8.3.